numba/tests/test_parfors.py

```diff
@@ -133,26 +135,34 @@ def copy_args(*args):
                 elif isinstance(x, numbers.Number):
                     new_args.append(x)
                 elif isinstance(x, tuple):
-                    new_args.append(x)
+                    new_args.append(copy.deepcopy(x))
                 elif isinstance(x, list):
                     new_args.append(x[:])
                 else:
                     raise ValueError('Unsupported argument type encountered')
             return tuple(new_args)
 
         # python result
-        py_expected = pyfunc(*copy_args(*args))
+        py_args = copy_args(*args)
+        py_expected = pyfunc(*py_args)
 
         # njit result
-        njit_output = cfunc.entry_point(*copy_args(*args))
+        njit_args = copy_args(*args)
+        njit_output = cfunc.entry_point(*njit_args)
 
         # parfor result
-        parfor_output = cpfunc.entry_point(*copy_args(*args))
-
-        np.testing.assert_almost_equal(njit_output, py_expected, **kwargs)
-        np.testing.assert_almost_equal(parfor_output, py_expected, **kwargs)
+        parfor_args = copy_args(*args)
+        parfor_output = cpfunc.entry_point(*parfor_args)
 
-        self.assertEqual(type(njit_output), type(parfor_output))
+        if check_args_for_equality is None:
+            np.testing.assert_almost_equal(njit_output, py_expected, **kwargs)
+            np.testing.assert_almost_equal(parfor_output, py_expected, **kwargs)
+            self.assertEqual(type(njit_output), type(parfor_output))
+        else:
+            for pyarg, njitarg, parforarg, argcomp in zip(
```

Comment:
I think it'd be good to put a check in here to make sure that `check_args_for_equality` is the same size as e.g. `pyarg` so that the iteration doesn't stop short because insufficient checking functions were supplied. Might also be a good idea to have a "single checking function" option?

Reply:
@stuartarchibald I intentionally did it this way.  That way, if you have multiple arguments but only need to check the first one then you only need to give one comparison function.  When writing tests, you can then put the ones you need to compare at the beginning but it isn't a disaster if you don't because you can provide a comparator that always returns True.  So, we need to come to agreement on the right approach.  If mine is good then I think the PR is fine, maybe some more documentation about the feature?  If yours is selected then add that check and documentation.  Let me know what you think.

Reply:
I see, thanks for explaining. As tests are so complicated I'm inclined to go with explicit size match with check and document that as such opposed to relying on test authors recalling/reading the docs and reviewers catching systematic errors. Perhaps a null comparator can be added to the module so it is trivial to just add as needed? Thoughts?

Reply:
@stuartarchibald I implemented your suggestion but in the process of pushing the commit it rejected it due to rejecting fast-forwards.  So, I did a git pull and now I'm seeing some testing error about the number of parfors in some test.  This code was passing CI before the merge?  Do you think there's some interaction causing this or did some error creep in somehow in the merge?

Reply:
This PR will be merged against current master as part of the CI test strategy, I guess there's some interaction in that that is causing the problem.
```
$ git rev-parse HEAD # THIS IS THE HEAD OF THIS BRANCH
d9b60690ecf7c3aa078f13dd2cf1b31ce9649655

$ ./runtests.py numba.tests.test_parfors.TestParfors.test_logistic_regression
.
----------------------------------------------------------------------
Ran 1 test in 0.675s

OK

$ git merge master
Auto-merging numba/tests/test_parfors.py
Auto-merging numba/core/ir_utils.py
Merge made by the 'recursive' strategy.
 docs/source/conf.py                       |   2 +-
 docs/source/reference/envvars.rst         |   4 ++
 docs/source/user/installing.rst           |  56 +++++++++++++----
 numba/_helperlib.c                        | 266 +++++++++++++++++++++++++++++++++++++-----------------------------------------
 numba/core/codegen.py                     |   2 +-
 numba/core/errors.py                      |  12 ++--
 numba/core/inline_closurecall.py          |   2 +
 numba/core/ir_utils.py                    |   4 +-
 numba/core/object_mode_passes.py          |   2 +-
 numba/core/typeconv/typeconv.py           |   2 +-
 numba/core/typed_passes.py                |   2 +-
 numba/core/typeinfer.py                   |   2 +-
 numba/core/types/containers.py            |   5 +-
 numba/core/types/functions.py             |  21 ++++++-
 numba/core/typing/arraydecl.py            |   2 +
 numba/core/typing/npydecl.py              |   4 ++
 numba/core/typing/templates.py            |  18 ++++++
 numba/cpython/tupleobj.py                 |  18 ++++++
 numba/cuda/envvars.py                     |   2 +-
 numba/experimental/jitclass/decorators.py |   2 +-
 numba/misc/gdb_hook.py                    |   2 +-
 numba/misc/literal.py                     |   3 +
 numba/np/ufunc/parallel.py                |  49 +++++++++------
 numba/parfors/array_analysis.py           |   2 +
 numba/parfors/parfor.py                   |  61 +++++++++++-------
 numba/tests/test_comprehension.py         |  10 +++
 numba/tests/test_deprecations.py          |   4 +-
 numba/tests/test_parallel_backend.py      |  58 +++++++++++------
 numba/tests/test_parfors.py               |  40 ++++++++++++
 numba/tests/test_tuples.py                |  13 +++-
 setup.py                                  |  77 ++++++++++++-----------
 31 files changed, 474 insertions(+), 273 deletions(-)

$ ./runtests.py numba.tests.test_parfors.TestParfors.test_logistic_regression
F
======================================================================
FAIL: test_logistic_regression (numba.tests.test_parfors.TestParfors)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "<path>/numba/numba/tests/test_parfors.py", line 605, in test_logistic_regression
    self.assertTrue(countParfors(lr_impl, args) == 2)
AssertionError: False is not true

----------------------------------------------------------------------
Ran 1 test in 0.457s

```